In [1]:
import pandas as pd

dataset = pd.read_csv('../data/Shakespeare_data.csv')

I hypothesized that certain characters might say things more fancily or in a more flowery way than other characters. I'm using the python library textstat to calculate the Flesch-Kinkaid Reading Ease rating for a particular PlayerLine. This rates a given text on a scale of 1-121.22, with higher scores meaning the text is less difficult to read, and lower scores meaning the text is more difficult to read. This might allow for a classifier to be more accurate, as scores might tend to cluster differently depending on the character speaking the line.

In [2]:
import textstat

dataset['dale_chall_readability'] = dataset.apply(lambda row: textstat.dale_chall_readability_score(row['PlayerLine']), axis=1)

dataset

Dataline            Play  PlayerLinenumber ActSceneLine  \
0              1        Henry IV               NaN          NaN   
1              2        Henry IV               NaN          NaN   
2              3        Henry IV               NaN          NaN   
3              4        Henry IV               1.0        1.1.1   
4              5        Henry IV               1.0        1.1.2   
5              6        Henry IV               1.0        1.1.3   
6              7        Henry IV               1.0        1.1.4   
7              8        Henry IV               1.0        1.1.5   
8              9        Henry IV               1.0        1.1.6   
9             10        Henry IV               1.0        1.1.7   
10            11        Henry IV               1.0        1.1.8   
11            12        Henry IV               1.0        1.1.9   
12            13        Henry IV               1.0       1.1.10   
13            14        Henry IV               1.0       1.1.11   
14            15        Henry IV               1.0       1.1.12   
15            16        Henry IV               1.0       1.1.13   
16            17        Henry IV               1.0       1.1.14   
17            18        Henry IV               1.0       1.1.15   
18            19        Henry IV               1.0       1.1.16   
19            20        Henry IV               1.0       1.1.17   
20            21        Henry IV               1.0       1.1.18   
21            22        Henry IV               1.0       1.1.19   
22            23        Henry IV               1.0       1.1.20   
23            24        Henry IV               1.0       1.1.21   
24            25        Henry IV               1.0       1.1.22   
25            26        Henry IV               1.0       1.1.23   
26            27        Henry IV               1.0       1.1.24   
27            28        Henry IV               1.0       1.1.25   
28            29        Henry IV               1.0       1.1.26   
29            30        Henry IV               1.0       1.1.27   
...          ...             ...               ...          ...   
111366    111367  A Winters Tale              37.0      5.3.155   
111367    111368  A Winters Tale              37.0      5.3.156   
111368    111369  A Winters Tale              37.0      5.3.157   
111369    111370  A Winters Tale              37.0      5.3.158   
111370    111371  A Winters Tale              37.0      5.3.159   
111371    111372  A Winters Tale              37.0      5.3.160   
111372    111373  A Winters Tale              37.0      5.3.161   
111373    111374  A Winters Tale              37.0      5.3.162   
111374    111375  A Winters Tale              38.0      5.3.163   
111375    111376  A Winters Tale              38.0      5.3.164   
111376    111377  A Winters Tale              38.0      5.3.165   
111377    111378  A Winters Tale              38.0      5.3.166   
111378    111379  A Winters Tale              38.0      5.3.167   
111379    111380  A Winters Tale              38.0      5.3.168   
111380    111381  A Winters Tale              38.0      5.3.169   
111381    111382  A Winters Tale              38.0      5.3.170   
111382    111383  A Winters Tale              38.0      5.3.171   
111383    111384  A Winters Tale              38.0      5.3.172   
111384    111385  A Winters Tale              38.0      5.3.173   
111385    111386  A Winters Tale              38.0      5.3.174   
111386    111387  A Winters Tale              38.0      5.3.175   
111387    111388  A Winters Tale              38.0      5.3.176   
111388    111389  A Winters Tale              38.0      5.3.177   
111389    111390  A Winters Tale              38.0      5.3.178   
111390    111391  A Winters Tale              38.0      5.3.179   
111391    111392  A Winters Tale              38.0      5.3.180   
111392    111393  A Winters Tale              38.0      5.3.181   
111393    111394  A Winters Tale              38.0     

Another interesting data point to look at might be the average length of words used in the line, as certain characters might have a propensity to use longer words.

In [6]:
def avg_word_length(text):
    text = [word.replace(".", "").replace(",", "").replace(";", "").replace(":","") for word in text.split()]
    text_lens = sum([len(word) for word in text])
    return text_lens / len(text)

dataset['avg_word_length'] = dataset.apply(lambda row: avg_word_length(row['PlayerLine']), axis=1)
dataset

C:\Users\jam60\Anaconda3\lib\site-packages\ipykernel_launcher.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  


Dataline            Play  PlayerLinenumber ActSceneLine  \
3              4        Henry IV               1.0        1.1.1   
4              5        Henry IV               1.0        1.1.2   
5              6        Henry IV               1.0        1.1.3   
6              7        Henry IV               1.0        1.1.4   
7              8        Henry IV               1.0        1.1.5   
8              9        Henry IV               1.0        1.1.6   
9             10        Henry IV               1.0        1.1.7   
10            11        Henry IV               1.0        1.1.8   
11            12        Henry IV               1.0        1.1.9   
12            13        Henry IV               1.0       1.1.10   
13            14        Henry IV               1.0       1.1.11   
14            15        Henry IV               1.0       1.1.12   
15            16        Henry IV               1.0       1.1.13   
16            17        Henry IV               1.0       1.1.14   
17            18        Henry IV               1.0       1.1.15   
18            19        Henry IV               1.0       1.1.16   
19            20        Henry IV               1.0       1.1.17   
20            21        Henry IV               1.0       1.1.18   
21            22        Henry IV               1.0       1.1.19   
22            23        Henry IV               1.0       1.1.20   
23            24        Henry IV               1.0       1.1.21   
24            25        Henry IV               1.0       1.1.22   
25            26        Henry IV               1.0       1.1.23   
26            27        Henry IV               1.0       1.1.24   
27            28        Henry IV               1.0       1.1.25   
28            29        Henry IV               1.0       1.1.26   
29            30        Henry IV               1.0       1.1.27   
30            31        Henry IV               1.0       1.1.28   
31            32        Henry IV               1.0       1.1.29   
32            33        Henry IV               1.0       1.1.30   
...          ...             ...               ...          ...   
111365    111366  A Winters Tale              36.0      5.3.154   
111366    111367  A Winters Tale              37.0      5.3.155   
111367    111368  A Winters Tale              37.0      5.3.156   
111368    111369  A Winters Tale              37.0      5.3.157   
111369    111370  A Winters Tale              37.0      5.3.158   
111370    111371  A Winters Tale              37.0      5.3.159   
111371    111372  A Winters Tale              37.0      5.3.160   
111372    111373  A Winters Tale              37.0      5.3.161   
111373    111374  A Winters Tale              37.0      5.3.162   
111374    111375  A Winters Tale              38.0      5.3.163   
111375    111376  A Winters Tale              38.0      5.3.164   
111376    111377  A Winters Tale              38.0      5.3.165   
111377    111378  A Winters Tale              38.0      5.3.166   
111378    111379  A Winters Tale              38.0      5.3.167   
111379    111380  A Winters Tale              38.0      5.3.168   
111380    111381  A Winters Tale              38.0      5.3.169   
111381    111382  A Winters Tale              38.0      5.3.170   
111382    111383  A Winters Tale              38.0      5.3.171   
111383    111384  A Winters Tale              38.0      5.3.172   
111384    111385  A Winters Tale              38.0      5.3.173   
111385    111386  A Winters Tale              38.0      5.3.174   
111386    111387  A Winters Tale              38.0      5.3.175   
111387    111388  A Winters Tale              38.0      5.3.176   
111388    111389  A Winters Tale              38.0      5.3.177   
111389    111390  A Winters Tale              38.0      5.3.178   
111390    111391  A Winters Tale              38.0      5.3.179   
111391    111392  A Winters Tale              38.0      5.3.180   
111392    111393  A Winters Tale              38.0     

One thing we might want to do first: remove the rows that aren't said by characters (but are stage directions or something else). Let's do that:

In [7]:
dataset = dataset[(dataset.PlayerLinenumber.notnull()) & (dataset.ActSceneLine.notnull()) & (dataset.Player.notnull())]
dataset

Dataline            Play  PlayerLinenumber ActSceneLine  \
3              4        Henry IV               1.0        1.1.1   
4              5        Henry IV               1.0        1.1.2   
5              6        Henry IV               1.0        1.1.3   
6              7        Henry IV               1.0        1.1.4   
7              8        Henry IV               1.0        1.1.5   
8              9        Henry IV               1.0        1.1.6   
9             10        Henry IV               1.0        1.1.7   
10            11        Henry IV               1.0        1.1.8   
11            12        Henry IV               1.0        1.1.9   
12            13        Henry IV               1.0       1.1.10   
13            14        Henry IV               1.0       1.1.11   
14            15        Henry IV               1.0       1.1.12   
15            16        Henry IV               1.0       1.1.13   
16            17        Henry IV               1.0       1.1.14   
17            18        Henry IV               1.0       1.1.15   
18            19        Henry IV               1.0       1.1.16   
19            20        Henry IV               1.0       1.1.17   
20            21        Henry IV               1.0       1.1.18   
21            22        Henry IV               1.0       1.1.19   
22            23        Henry IV               1.0       1.1.20   
23            24        Henry IV               1.0       1.1.21   
24            25        Henry IV               1.0       1.1.22   
25            26        Henry IV               1.0       1.1.23   
26            27        Henry IV               1.0       1.1.24   
27            28        Henry IV               1.0       1.1.25   
28            29        Henry IV               1.0       1.1.26   
29            30        Henry IV               1.0       1.1.27   
30            31        Henry IV               1.0       1.1.28   
31            32        Henry IV               1.0       1.1.29   
32            33        Henry IV               1.0       1.1.30   
...          ...             ...               ...          ...   
111365    111366  A Winters Tale              36.0      5.3.154   
111366    111367  A Winters Tale              37.0      5.3.155   
111367    111368  A Winters Tale              37.0      5.3.156   
111368    111369  A Winters Tale              37.0      5.3.157   
111369    111370  A Winters Tale              37.0      5.3.158   
111370    111371  A Winters Tale              37.0      5.3.159   
111371    111372  A Winters Tale              37.0      5.3.160   
111372    111373  A Winters Tale              37.0      5.3.161   
111373    111374  A Winters Tale              37.0      5.3.162   
111374    111375  A Winters Tale              38.0      5.3.163   
111375    111376  A Winters Tale              38.0      5.3.164   
111376    111377  A Winters Tale              38.0      5.3.165   
111377    111378  A Winters Tale              38.0      5.3.166   
111378    111379  A Winters Tale              38.0      5.3.167   
111379    111380  A Winters Tale              38.0      5.3.168   
111380    111381  A Winters Tale              38.0      5.3.169   
111381    111382  A Winters Tale              38.0      5.3.170   
111382    111383  A Winters Tale              38.0      5.3.171   
111383    111384  A Winters Tale              38.0      5.3.172   
111384    111385  A Winters Tale              38.0      5.3.173   
111385    111386  A Winters Tale              38.0      5.3.174   
111386    111387  A Winters Tale              38.0      5.3.175   
111387    111388  A Winters Tale              38.0      5.3.176   
111388    111389  A Winters Tale              38.0      5.3.177   
111389    111390  A Winters Tale              38.0      5.3.178   
111390    111391  A Winters Tale              38.0      5.3.179   
111391    111392  A Winters Tale              38.0      5.3.180   
111392    111393  A Winters Tale              38.0     

Lets also convert as many columns as possible to float/integer values so we can use them as features in classification.

We can convert the plays to indices, and we can convert ActSceneLine into independent columns:

In [9]:
unique_plays = list(dataset.Play.unique())
def conv_to_index(play_name):
    return unique_plays.index(play_name)

dataset['Play'] = dataset['Play'].map(lambda x: conv_to_index(x))
dataset[['Act', 'Scene', 'Line']] = dataset['ActSceneLine'].str.split(pat='.', expand=True)
dataset
 

Dataline  Play  PlayerLinenumber ActSceneLine         Player  \
3              4     0               1.0        1.1.1  KING HENRY IV   
4              5     0               1.0        1.1.2  KING HENRY IV   
5              6     0               1.0        1.1.3  KING HENRY IV   
6              7     0               1.0        1.1.4  KING HENRY IV   
7              8     0               1.0        1.1.5  KING HENRY IV   
8              9     0               1.0        1.1.6  KING HENRY IV   
9             10     0               1.0        1.1.7  KING HENRY IV   
10            11     0               1.0        1.1.8  KING HENRY IV   
11            12     0               1.0        1.1.9  KING HENRY IV   
12            13     0               1.0       1.1.10  KING HENRY IV   
13            14     0               1.0       1.1.11  KING HENRY IV   
14            15     0               1.0       1.1.12  KING HENRY IV   
15            16     0               1.0       1.1.13  KING HENRY IV   
16            17     0               1.0       1.1.14  KING HENRY IV   
17            18     0               1.0       1.1.15  KING HENRY IV   
18            19     0               1.0       1.1.16  KING HENRY IV   
19            20     0               1.0       1.1.17  KING HENRY IV   
20            21     0               1.0       1.1.18  KING HENRY IV   
21            22     0               1.0       1.1.19  KING HENRY IV   
22            23     0               1.0       1.1.20  KING HENRY IV   
23            24     0               1.0       1.1.21  KING HENRY IV   
24            25     0               1.0       1.1.22  KING HENRY IV   
25            26     0               1.0       1.1.23  KING HENRY IV   
26            27     0               1.0       1.1.24  KING HENRY IV   
27            28     0               1.0       1.1.25  KING HENRY IV   
28            29     0               1.0       1.1.26  KING HENRY IV   
29            30     0               1.0       1.1.27  KING HENRY IV   
30            31     0               1.0       1.1.28  KING HENRY IV   
31            32     0               1.0       1.1.29  KING HENRY IV   
32            33     0               1.0       1.1.30  KING HENRY IV   
...          ...   ...               ...          ...            ...   
111365    111366    35              36.0      5.3.154       HERMIONE   
111366    111367    35              37.0      5.3.155        PAULINA   
111367    111368    35              37.0      5.3.156        PAULINA   
111368    111369    35              37.0      5.3.157        PAULINA   
111369    111370    35              37.0      5.3.158        PAULINA   
111370    111371    35              37.0      5.3.159        PAULINA   
111371    111372    35              37.0      5.3.160        PAULINA   
111372    111373    35              37.0      5.3.161        PAULINA   
111373    111374    35              37.0      5.3.162        PAULINA   
111374    111375    35              38.0      5.3.163        LEONTES   
111375    111376    35              38.0      5.3.164        LEONTES   
111376    111377    35              38.0      5.3.165        LEONTES   
111377    111378    35              38.0      5.3.166        LEONTES   
111378    111379    35              38.0      5.3.167        LEONTES   
111379    111380    35              38.0      5.3.168        LEONTES   
111380    111381    35              38.0      5.3.169        LEONTES   
111381    111382    35              38.0      5.3.170        LEONTES   
111382    111383    35              38.0      5.3.171        LEONTES   
111383    111384    35              38.0      5.3.172        LEONTES   
111384    111385    35              38.0      5.3.173        LEONTES   
111385    111386    35              38.0      5.3.174        LEONTES   
111386    111387    35              38.0      5.3.175        LEONTES   
111387    111388    35              38.0      5.3.176        LEONTES   
111388    111389    35              38.0      5.

For this dataset we will attempt to classify using a Random Forest Classifier. 

In [10]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
import numpy as np

dataset.columns
x_dataset = dataset[['Play', 'PlayerLinenumber', 'dale_chall_readability', 'avg_word_length', 'Act', 'Scene', 'Line']]
y_dataset = dataset[['Player']]

X = x_dataset.to_numpy()
Y = y_dataset.to_numpy()

X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2)

rfc = RandomForestClassifier()
rfc.fit(X_train, Y_train)
predictions = rfc.predict(X_test)
accuracy_score = accuracy_score(Y_test, predictions)
accuracy_score



C:\Users\jam60\Anaconda3\lib\site-packages\sklearn\ensemble\forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)
C:\Users\jam60\Anaconda3\lib\site-packages\ipykernel_launcher.py:16: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  app.launch_new_instance()


0.6920736056297846

My accuracy score was 0.692, which is not horrible given that we trained on some peculiar data in average word length and Dale-Chall Score.